### ISSUE TITLE
Error: module 'rope.base.ast' has no attribute Str

### ISSUE THREAD

**Issue by iogf** [python-rope/rope#279]
That exception occurs when it try to rename any class in :

https://github.com/vyapp/vy/blob/master/vyapp/areavi.py

The plugin for vy is:

https://github.com/vyapp/vy/blob/master/vyapp/plugins/rope.py

The way of how it works consists of passing one of the offsets of the object name thats gonna be renamed altogether with the new name as it is explained in the docs.

However it seem to work for smaller projects but failing with bigger ones.

I'm using python3.8 on an arch system.

The full excepltion below:

~~~
[tau@archlinux untwisted-code]$ git checkout *
Updated 1 path from the index
[tau@archlinux untwisted-code]$ cd ..
[tau@archlinux projects]$ vy vy-code/vyapp/areavi.py -v
Loading vyrc...
Project - Setting sentinels =  ('.git', '.svn', '.hg', '._')
Quick Search - Setting confs =  {'background': 'yellow', 'foreground': 'black'}
Find - Setting confs =  {'background': 'green', 'foreground': 'white'}
Sniper - Setting dirs = /home/tau/projects
Sniper - Setting ag path =  ag
Fstmt - Setting ag path =  ag
Snakerr - Setting Pyflakes path =  pyflakes
Snakerr - Setting Mypy path =  mypy
Deadcode - Setting Vulture path =  vulture
Project - Setting project path =  /home/tau/projects/vy-code
Exception in Tkinter callback
Traceback (most recent call last):
  File "/usr/lib/python3.8/tkinter/__init__.py", line 1883, in __call__
    return self.func(*args)
  File "/usr/lib/python3.8/site-packages/vyapp/tools.py", line 48, in shell
    return handle(*args, **kwargs)
  File "/usr/lib/python3.8/site-packages/vyapp/plugins/rope.py", line 119, in get_rename_data
    self.rename(ask.data)
  File "/usr/lib/python3.8/site-packages/vyapp/plugins/rope.py", line 125, in rename
    project = Project(path)
  File "/usr/lib/python3.8/site-packages/rope/base/project.py", line 223, in __init__
    self._init_prefs(prefs)
  File "/usr/lib/python3.8/site-packages/rope/base/project.py", line 273, in _init_prefs
    self._init_ropefolder()
  File "/usr/lib/python3.8/site-packages/rope/base/project.py", line 249, in _init_ropefolder
    config.write(self._default_config())
  File "/usr/lib/python3.8/site-packages/rope/base/resources.py", line 128, in write
    self._perform_change(change.ChangeContents(self, contents),
  File "/usr/lib/python3.8/site-packages/rope/base/resources.py", line 99, in _perform_change
    self.project.do(changes)
  File "/usr/lib/python3.8/site-packages/rope/base/project.py", line 113, in do
    self.history.do(changes, task_handle=task_handle)
  File "/usr/lib/python3.8/site-packages/rope/base/history.py", line 36, in do
    changes.do(change.create_job_set(task_handle, changes))
  File "/usr/lib/python3.8/site-packages/rope/base/change.py", line 65, in do
    change.do(job_set)
  File "/usr/lib/python3.8/site-packages/rope/base/change.py", line 126, in call
    function(self)
  File "/usr/lib/python3.8/site-packages/rope/base/change.py", line 150, in do
    self._operations.write_file(self.resource, self.new_contents)
  File "/usr/lib/python3.8/site-packages/rope/base/change.py", line 335, in write_file
    observer.resource_changed(resource)
  File "/usr/lib/python3.8/site-packages/rope/base/resourceobserver.py", line 30, in resource_changed
    self.changed(resource)
  File "/usr/lib/python3.8/site-packages/rope/base/pycore.py", line 65, in _file_changed_for_soa
    perform_soa_on_changed_scopes(self.project, resource, old_contents)
  File "/usr/lib/python3.8/site-packages/rope/base/pycore.py", line 302, in perform_soa_on_changed_scopes
    pycore.analyze_module(resource, should_analyze, search_subscopes)
  File "/usr/lib/python3.8/site-packages/rope/base/pycore.py", line 209, in analyze_module
    rope.base.oi.soa.analyze_module(
  File "/usr/lib/python3.8/site-packages/rope/base/oi/soa.py", line 15, in analyze_module
    _analyze_node(pycore, pymodule, should_analyze,
  File "/usr/lib/python3.8/site-packages/rope/base/oi/soa.py", line 23, in _analyze_node
    _analyze_node(pycore, scope.pyobject, should_analyze,
  File "/usr/lib/python3.8/site-packages/rope/base/oi/soa.py", line 38, in _analyze_node
    rope.base.ast.walk(child, visitor)
  File "/usr/lib/python3.8/site-packages/rope/base/ast.py", line 39, in walk
    return method(node)
  File "/usr/lib/python3.8/site-packages/rope/base/oi/soa.py", line 122, in _Assign
    pyobject = instance.get_object()
  File "/usr/lib/python3.8/site-packages/rope/base/pynamesdef.py", line 42, in get_object
    result = self.pyfunction.get_parameter(self.index)
  File "/usr/lib/python3.8/site-packages/rope/base/pyobjectsdef.py", line 69, in get_parameter
    if index < len(self.parameter_pyobjects.get()):
  File "/usr/lib/python3.8/site-packages/rope/base/utils/__init__.py", line 29, in newfunc
    return func(self, *args, **kwds)
  File "/usr/lib/python3.8/site-packages/rope/base/pynames.py", line 188, in get
    self.set(self.get_inferred(*args, **kwds))
  File "/usr/lib/python3.8/site-packages/rope/base/pyobjectsdef.py", line 40, in _infer_parameters
    pyobjects = rope.base.oi.soi.infer_parameter_objects(self)
  File "/usr/lib/python3.8/site-packages/rope/base/utils/__init__.py", line 41, in newfunc
    return func(*args, **kwds)
  File "/usr/lib/python3.8/site-packages/rope/base/oi/soi.py", line 47, in infer_parameter_objects
    result = _parameter_objects(pyfunction)
  File "/usr/lib/python3.8/site-packages/rope/base/oi/soi.py", line 130, in _parameter_objects
    type_ = hint_param(pyobject, name)
  File "/usr/lib/python3.8/site-packages/rope/base/oi/type_hinting/providers/inheritance.py", line 21, in __call__
    result = self._delegate(superfunc, param_name)
  File "/usr/lib/python3.8/site-packages/rope/base/oi/type_hinting/providers/composite.py", line 19, in __call__
    result = delegate(pyfunc, param_name)
  File "/usr/lib/python3.8/site-packages/rope/base/oi/type_hinting/providers/docstrings.py", line 46, in __call__
    type_strs = self._parse_docstring(pyfunc.get_doc(), param_name)
  File "/usr/lib/python3.8/site-packages/rope/base/pyobjects.py", line 224, in get_doc
    isinstance(expr.value, ast.Str):
AttributeError: module 'rope.base.ast' has no attribute 'Str'
~~~

**Comment by caga** [comment]
any progress?

**Comment by mcepl** [comment]
Patches welcome!

**Comment by mcepl** [comment]
> any progress?

Actually, could you make me some short self-contained test script presenting this issue, please? We could have at least test to start with.

**Comment by mcepl** [comment]
@iogf Are you able to test this issue with the pull request #290, please?

**Comment by iogf** [comment]
I will test then get back to you soon.

**Comment by FranciscoDA** [comment]
@mcepl I was having the same issue and the package from your branch fixed it for me. Thanks

**Comment by mcepl** [comment]
Perfect, thank you for testing.

**Comment by 0alpha** [comment]
Do you have an estimate for when the next release will come out?

**Comment by mcepl** [comment]
@0alpha, @iogf Can you reproduce with the current master (d46aa2dcf34c6baf57b37930f5db7dc962b770b6), please?

**Comment by mkorvas** [comment]
Sorry I dropped out of discussions of these issues.

I can confirm that (in my test case that's far from minimal)
- I see the error message about "no attribute 'Str'" on c6574094 (a commit currently pointed to by vim's python-mode plugin)
- the error is gone when I cherry-pick 644fb2f9 (the only commit of https://github.com/python-rope/rope/pull/290)
- the error is gone when I check out the `master` branch (d46aa2d).

**Comment by mcepl** [comment]
Thank you.

**Comment by mcepl** [comment]
> Do you have an estimate for when the next release will come out?

0alpha, right now.

**Comment by shaypal5** [comment]
As of August 2020, the fixed code cannot be found on master, and the commit itself seems to have been deleted from history:

https://github.com/python-rope/rope/commits/master
https://github.com/python-rope/rope/commits/master/rope/base/pyobjects.py

**Comment by estyrke** [comment]
@shaypal5 But is the issue still present? Rope 0.18.0 fixes the issue for me (relative to rope 0.16.0 where I found the issue).
